WordPress/src/main/java/org/wordpress/android/ui/mysite/cards/dashboard/CardsSource.kt

```diff
@@ -0,0 +1,82 @@
+import androidx.lifecycle.MutableLiveData
+import kotlinx.coroutines.CoroutineDispatcher
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.flow.collect
+import kotlinx.coroutines.launch
+import org.wordpress.android.fluxc.model.SiteModel
+import org.wordpress.android.fluxc.store.dashboard.CardsStore
+import org.wordpress.android.modules.BG_THREAD
+import org.wordpress.android.ui.mysite.MySiteSource.MySiteRefreshSource
+import org.wordpress.android.ui.mysite.MySiteUiState.PartialState.CardsUpdate
+import org.wordpress.android.ui.mysite.SelectedSiteRepository
+import javax.inject.Inject
+import javax.inject.Named
+import javax.inject.Singleton
+
+@Singleton
+class CardsSource @Inject constructor(
+    private val selectedSiteRepository: SelectedSiteRepository,
+    private val cardsStore: CardsStore,
+    @param:Named(BG_THREAD) private val bgDispatcher: CoroutineDispatcher
+) : MySiteRefreshSource<CardsUpdate> {
+    override val refresh: MutableLiveData<Boolean> = MutableLiveData<Boolean>(false)
+
+    override fun build(coroutineScope: CoroutineScope, siteLocalId: Int): LiveData<CardsUpdate> {
+        val result = MediatorLiveData<CardsUpdate>()
+        result.getData(coroutineScope, siteLocalId)
+        result.refreshData(coroutineScope, siteLocalId)
+        result.addSource(refresh) { result.refreshData(coroutineScope, siteLocalId, refresh.value) }
+        return result
+    }
+
+    private fun MediatorLiveData<CardsUpdate>.getData(
+        coroutineScope: CoroutineScope,
+        siteLocalId: Int
+    ) {
+        val selectedSite = selectedSiteRepository.getSelectedSite()
+        if (selectedSite != null && selectedSite.id == siteLocalId) {
+            coroutineScope.launch(bgDispatcher) {
+                cardsStore.getCards(selectedSite).collect { result ->
+                    postState(CardsUpdate(result))
+                }
+            }
+        }
+    }
+
+    private fun MediatorLiveData<CardsUpdate>.refreshData(
+        coroutineScope: CoroutineScope,
+        siteLocalId: Int,
+        isRefresh: Boolean? = null
+    ) {
+        when (isRefresh) {
+            null, true -> refreshData(coroutineScope, siteLocalId)
+            else -> Unit // Do nothing
+        }
+    }
+
+    private fun MediatorLiveData<CardsUpdate>.refreshData(
+        coroutineScope: CoroutineScope,
+        siteLocalId: Int
+    ) {
+        val selectedSite = selectedSiteRepository.getSelectedSite()
+        if (selectedSite != null && selectedSite.id == siteLocalId) {
+            fetchCardsAndPostErrorIfAvailable(coroutineScope, selectedSite)
+        }
+    }
+
+    private fun MediatorLiveData<CardsUpdate>.fetchCardsAndPostErrorIfAvailable(
+        coroutineScope: CoroutineScope,
+        selectedSite: SiteModel
+    ) {
+        coroutineScope.launch(bgDispatcher) {
+            val result = cardsStore.fetchCards(selectedSite)
+            result.error?.let {
+                postState(CardsUpdate(result))
```

Comment:
Similar ❓  here ... if there is an error, we still need some type of update to flip the refresh flag to false? Thoughts?

Reply:
👋 @zwarm !

Thanks for the question. Actually, what you see there is that if there is an error, and only an error we are posting the state, thus you are seeing the `result.error?.let { ... }`. Then, the flow above will capture the rest.

Or, am I not understanding your question well?

Reply:
Yes, I thought about this at 2AM ... this is only the error part. Thanks for clarifying.